### PR TITLE
Update AzureML Workspace template to support all scenarios

### PR DESCRIPTION
**Pull request by guanyu-240**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

* Add private link support (Auto-approval and manual approval)
* Support storage/kv/acr behind VNET
* Limit private endpoint deployment in selected regions only

### CODE REVIEW

**Comment by guanyu-240** [comment]
@bmoore-msft Could you have someone take a look at this PR? Thank you!

File: 101-machine-learning-create/azuredeploy.json [hunk 1]
```diff
@@ -2,257 +2,641 @@
...
+      "allowedValues": [
+        "AutoApproval",
+        "ManualApproval",
+        "none"
+      ]
+    }
+
+  },
+  "variables": {
+    "storageAccountName": "[if(empty(parameters('storageAccountName')), concat('sa',uniqueString(resourceGroup().id, parameters('location'))), parameters('storageAccountName'))]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:277]: revert this change (remove all the if() stmts) and use the defaultValues on the parameters

File: 101-machine-learning-create/azuredeploy.json [hunk 2]
```diff
@@ -2,257 +2,641 @@
...
+    "subnetName": "[if(empty(parameters('subnetName')), concat('sn',uniqueString(resourceGroup().id, parameters('location'))), parameters('subnetName'))]",
+    "tenantId": "[subscription().tenantId]",
+    "storageAccount": "[resourceId(parameters('storageAccountResourceGroupName'), 'Microsoft.Storage/storageAccounts', variables('storageAccountName'))]",
+    "keyVault": "[resourceId(parameters('keyVaultResourceGroupName'), 'Microsoft.KeyVault/vaults', variables('keyVaultName'))]",
+    "containerRegistry": "[resourceId(parameters('containerRegistryResourceGroupName'), 'Microsoft.ContainerRegistry/registries', variables('containerRegistryName'))]",
+    "applicationInsights": "[resourceId(parameters('applicationInsightsResourceGroupName'), 'Microsoft.Insights/components', variables('applicationInsightsName'))]",
+    "vnet": "[resourceId(parameters('vnetResourceGroupName'), 'Microsoft.Network/virtualNetworks', variables('vnetName'))]",
+    "subnet": "[resourceId(parameters('vnetResourceGroupName'), 'Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('subnetName'))]",
+    "privateDnsGuid": "[guid(resourceGroup().id, deployment().name)]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:290]: use an idempotent unique value

File: 101-machine-learning-create/azuredeploy.json [hunk 3]
```diff
@@ -2,257 +2,641 @@
...
+      "allowedValues": [
+        "true",
+        "false"
+      ]
+    },
+    "storageAccountResourceGroupName": {
+      "type": "string",
+      "defaultValue": "[resourceGroup().name]"
+    },
+    "keyVaultOption": {
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:90]: revert this change

File: 101-machine-learning-create/azuredeploy.json [hunk 4]
```diff
@@ -2,257 +2,641 @@
...
+        "new",
+        "existing"
+      ],
+      "metadata": {
+        "description": "Determines whether or not a new storage should be provisioned."
+      }
+    },
+    "storageAccountName": {
+      "type": "string",
+      "defaultValue": ""
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:62]: revert this change

File: 101-machine-learning-create/azuredeploy.json [hunk 5]
```diff
@@ -2,257 +2,641 @@
...
+      "allowedValues": [
+        "true",
+        "false"
+      ]
+    },
+    "keyVaultResourceGroupName": {
+      "type": "string",
+      "defaultValue": "[resourceGroup().name]"
+    },
+    "applicationInsightsOption": {
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:117]: revert this change

File: 101-machine-learning-create/azuredeploy.json [hunk 6]
```diff
@@ -2,257 +2,641 @@
...
+        "new",
+        "existing"
+      ],
+      "metadata": {
+        "description": "Determines whether or not new ApplicationInsights should be provisioned."
+      }
+    },
+    "applicationInsightsName": {
+      "type": "string",
+      "defaultValue": ""
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:130]: revert this change

File: 101-machine-learning-create/azuredeploy.json [hunk 7]
```diff
@@ -2,257 +2,641 @@
...
+        "existing",
+        "none"
+      ],
+      "metadata": {
+        "description": "Determines whether or not a new container registry should be provisioned."
+      }
+    },
+    "containerRegistryName": {
+      "type": "string",
+      "defaultValue": "",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:150]: add a valid defaultValue

File: 101-machine-learning-create/azuredeploy.json [hunk 8]
```diff
@@ -2,257 +2,641 @@
...
+      "defaultValue": "none",
+      "allowedValues": [
+        "new",
+        "existing",
+        "none"
+      ]
+    },
+    "vnetName": {
+      "type": "string",
+      "defaultValue": ""
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:187]: add a valid defaultValue

File: 101-machine-learning-create/azuredeploy.json [hunk 9]
```diff
@@ -2,257 +2,641 @@
...
+    "vnetOption": {
+      "type": "string",
+      "defaultValue": "none",
+      "allowedValues": [
+        "new",
+        "existing",
+        "none"
+      ]
+    },
+    "vnetName": {
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:185]: add a metadata description to all the params

File: 101-machine-learning-create/azuredeploy.json [hunk 10]
```diff
@@ -2,257 +2,641 @@
...
+      "defaultValue": [
+        "10.0.0.0/16"
+      ],
+      "metadata": {
+        "description": "Address prefix of the virtual network"
+      }
+    },
+    "subnetOption": {
+      "type": "string",
+      "defaultValue": "none",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:204]: is there a use case for provisioning a vnet with no subnet?

File: 101-machine-learning-create/azuredeploy.json [hunk 11]
```diff
@@ -2,257 +2,641 @@
...
+      "allowedValues": [
+        "new",
+        "existing",
+        "none"
+      ]
+    },
+    "subnetName": {
+      "type": "string",
+      "defaultValue": ""
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:213]: add a valid defaultValue

File: 101-machine-learning-create/azuredeploy.json [hunk 12]
```diff
@@ -2,257 +2,641 @@
...
+    "applicationInsights": "[resourceId(parameters('applicationInsightsResourceGroupName'), 'Microsoft.Insights/components', variables('applicationInsightsName'))]",
+    "vnet": "[resourceId(parameters('vnetResourceGroupName'), 'Microsoft.Network/virtualNetworks', variables('vnetName'))]",
+    "subnet": "[resourceId(parameters('vnetResourceGroupName'), 'Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('subnetName'))]",
+    "privateDnsGuid": "[guid(resourceGroup().id, deployment().name)]",
+    "locationsPEAvailable": [
+      "southcentralus",
+      "eastus",
+      "westus2"
+    ],
+    "isLocationValid": "[or(equals(parameters('privateEndpointType'), 'none'), contains(variables('locationsPEAvailable'), parameters('location')))]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:296]: remove this variable (you'll need to remove the pattern that uses below it as well)

File: 101-machine-learning-create/azuredeploy.json [hunk 13]
```diff
@@ -2,257 +2,641 @@
...
+    },
+    "defaultPEConnections": "[array(variables('privateEndpointSettings'))]"
+  },
+  "resources": [    
+    {
+      "condition": "[or(not(equals(parameters('vnetOption'), 'none')), not(equals(parameters('privateEndpointType'), 'none')))]",
+      "type": "Microsoft.Network/virtualNetworks",
+      "apiVersion": "2019-09-01",
+      "name": "[variables('vnetName')]",
+      "location": "[if(variables('isLocationValid'), parameters('location'), 'InvalidLocation')]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:329]: remove this pattern from the template

File: 101-machine-learning-create/azuredeploy.json [hunk 14]
```diff
@@ -2,257 +2,641 @@
...
+    {
+      "condition": "[equals(parameters('storageAccountBehindVNet'), 'true')]",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2017-05-10",
+      "name": "UpdateStorageAcls",
+      "dependsOn": [
+          "[variables('subnet')]",
+          "[variables('storageAccount')]"
+      ],
+      "resourceGroup": "[resourceGroup().name]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:397]: What's the intent here? is it to update an existing storage account?  If so, this needs to be set to the resourceGroup for that storage account, not the current RG.

File: 101-machine-learning-create/azuredeploy.json [hunk 15]
```diff
@@ -2,257 +2,641 @@
...
+          },
+          "keySource": "Microsoft.Storage"
+        },
+        "supportsHttpsTrafficOnly": true
+    }
+    },
+    {
+      "condition": "[equals(parameters('storageAccountBehindVNet'), 'true')]",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2017-05-10",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:391]: update apiVersion

File: 101-machine-learning-create/azuredeploy.json [hunk 16]
```diff
@@ -2,257 +2,641 @@
...
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "resources": [
+            {
+              "type": "Microsoft.Storage/storageAccounts",
+              "apiVersion": "2019-04-01",
+              "name": "[variables('storageAccountName')]",
+              "location": "[if(variables('isLocationValid'), parameters('location'), 'InvalidLocation')]",
+              "properties": "[union(reference(variables('storageAccount'), '2019-04-01'), variables('defaultNetworkAcls'))]"
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:409]: if this is updating the existing storage account it will replace current ACLs, which would be bad...  if necessary for this sample to work behind a vnet, you'll need to remove the option for an existing storage account

File: 101-machine-learning-create/azuredeploy.json [hunk 17]
```diff
@@ -2,257 +2,641 @@
...
+            "name": "standard",
+            "family": "A"
+        },
+        "accessPolicies": []
+      }
+    },
+    {
+      "condition": "[equals(parameters('keyVaultBehindVNet'), 'true')]",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2017-05-10",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:433]: update apiVersion

File: 101-machine-learning-create/azuredeploy.json [hunk 18]
```diff
@@ -2,257 +2,641 @@
...
+      "name": "UpdateKeyVaultAcls",
+      "dependsOn": [
+          "[variables('subnet')]",
+          "[variables('keyVault')]"
+      ],
+      "resourceGroup": "[resourceGroup().name]",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:439]: same as above

File: 101-machine-learning-create/azuredeploy.json [hunk 19]
```diff
@@ -2,257 +2,641 @@
...
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "resources": [
+            {
+              "type": "Microsoft.KeyVault/vaults",
+              "apiVersion": "2019-09-01",
+              "name": "[variables('keyVaultName')]",
+              "location": "[if(variables('isLocationValid'), parameters('location'), 'InvalidLocation')]",
+              "properties": "[union(reference(variables('keyVault'), '2019-09-01'), variables('defaultNetworkAcls'))]"
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:451]: same as with the storageAccount

File: 101-machine-learning-create/azuredeploy.json [hunk 20]
```diff
@@ -2,257 +2,641 @@
...
+          "resources": [
+            {
+              "type": "Microsoft.ContainerRegistry/registries",
+              "apiVersion": "2019-05-01",
+              "name": "[variables('containerRegistryName')]",
+              "location": "[if(variables('isLocationValid'), parameters('location'), 'InvalidLocation')]",
+              "sku": {
+                "name": "Premium"
+              },
+              "properties": "[if(equals(parameters('containerRegistryBehindVNet'), 'true'),
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:493]: same as above with your other existing resources...

File: 101-machine-learning-create/azuredeploy.json [hunk 21]
```diff
@@ -2,257 +2,641 @@
...
-                          "enabled": true
-                      },
-                      "file": {
-                          "enabled": true
-                      }
+    },
+    {
+      "condition": "[equals(parameters('privateEndpointType'), 'AutoApproval')]",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2017-05-10",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:566]: update apiVersion

File: 101-machine-learning-create/azuredeploy.json [hunk 22]
```diff
@@ -2,257 +2,641 @@
...
+              ],
+              "properties": {
+                "virtualNetwork": {
+                  "id": "[resourceId('Microsoft.Network/virtualNetworks', variables('vnetName'))]"
+                },
+                "registrationEnabled": false
+              }
+            },
+            {
+              "apiVersion": "2017-05-10",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:599]: update apiVersion

File: 101-machine-learning-create/azuredeploy.json [hunk 23]
```diff
@@ -2,257 +2,641 @@
...
+              "name": "[concat('EndpointDnsRecords-', variables('privateDnsGuid'))]",
+              "type": "Microsoft.Resources/deployments",
+              "dependsOn": [
+                "privatelink.api.azureml.ms"
+              ],
+              "properties": {
+                "mode": "Incremental",
+                "templatelink": {
+                  "contentVersion": "1.0.0.0",
+                  "uri": "https://network.hosting.portal.azure.net/network/Content/4.13.392.925/DeploymentTemplates/PrivateDnsForPrivateEndpoint.json"
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:609]: all files needed for deployment must be included as part of the sample/repo

File: 101-machine-learning-create/azuredeploy.json [hunk 24]
```diff
@@ -2,257 +2,641 @@
...
-              "applicationInsights": "[variables('applicationInsights')]",
-              "containerRegistry": "[if(empty(parameters('containerRegistry')), json('null'), parameters('containerRegistry'))]",
-              "adbWorkspace": "[if(empty(parameters('adbWorkspace')), json('null'), parameters('adbWorkspace'))]",
-              "encryption": {
-                  "status": "[parameters('encryption_status')]",
-                  "keyVaultProperties": {
-                      "keyVaultArmId": "[parameters('cmk_keyvault')]",
-                      "keyIdentifier": "[parameters('resource_cmk_uri')]"
+                  "privateEndpointNicResourceId": {
+                    "value": "[if(equals(parameters('privateEndpointType'), 'AutoApproval'),
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:616]: the false side of your if() stmt is not a valid expression

File: 101-machine-learning-create/azuredeploy.json [hunk 25]
```diff
@@ -2,257 +2,641 @@
...
+                    "value": "https://network.hosting.portal.azure.net/network/Content/4.13.392.925/DeploymentTemplates/PrivateDnsForPrivateEndpointNic.json"
+                  },
+                  "ipConfigRecordsTemplateUri": {
+                    "value": "https://network.hosting.portal.azure.net/network/Content/4.13.392.925/DeploymentTemplates/PrivateDnsForPrivateEndpointIpConfig.json"
+                  },
+                  "uniqueId": {
+                    "value": "[variables('privateDnsGuid')]"
+                  },
+                  "existingRecords": {
+                    "value": {}
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:631]: not sure I follow the intent with these deployments, since this is always empty...

File: 101-machine-learning-create/azuredeploy.json [hunk 26]
```diff
@@ -2,257 +2,641 @@
...
-              "hbiWorkspace": "[parameters('confidential_data')]"
-          }
-      }
+                }
+              }
+            }
+          ]
+        }
+      },
+      "resourceGroup": "[resourceGroup().name]"
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:639]: remove this property

File: 101-machine-learning-create/azuredeploy.json [hunk 27]
```diff
@@ -2,257 +2,641 @@
...
-                      "blob": {
-                          "enabled": true
-                      },
-                      "file": {
-                          "enabled": true
-                      }
+    },
+    {
+      "condition": "[equals(parameters('privateEndpointType'), 'AutoApproval')]",
+      "type": "Microsoft.Resources/deployments",
```
Comment (bmoore-msft) [inline 101-machine-learning-create/azuredeploy.json:565]: why is this nested?

File: 101-machine-learning-create/azuredeploy.json (hunk 12 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:296]: @bmoore-msft For this one. I need to limit the location to the selected locations only if private endpoint is not none. Any idea to handle this if not using this way?

File: 101-machine-learning-create/azuredeploy.json (hunk 24 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:616]: @bmoore-msft I tried. If not using this method, it will fail if the condition for the whole deployment is set to false

File: 101-machine-learning-create/azuredeploy.json (hunk 14 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:397]: @bmoore-msft Yes, update the ACL for new or existing storage account if behind Vnet

File: 101-machine-learning-create/azuredeploy.json (hunk 27 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:565]: @bmoore-msft Because it's a series of deployments based on private endpoint and all depends on the PE

File: 101-machine-learning-create/azuredeploy.json (hunk 24 above)
Comment (bmoore-msft) [reply 101-machine-learning-create/azuredeploy.json:616]: if you need the if() then use a valid expression on the false side of the statement... if it's a no-op, use something that indicates that (e.g. json('null'))

File: 101-machine-learning-create/azuredeploy.json (hunk 27 above)
Comment (bmoore-msft) [reply 101-machine-learning-create/azuredeploy.json:565]: You don't need nesting for dependencies... is there a problem you ran into?

File: 101-machine-learning-create/azuredeploy.json (hunk 16 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:409]: It's the requirement for updating existing storage account.

File: 101-machine-learning-create/azuredeploy.json (hunk 23 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json]: Refactored this section. Removed the files reference

File: 101-machine-learning-create/azuredeploy.json (hunk 25 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json]: Removed

File: 101-machine-learning-create/azuredeploy.json (hunk 16 above)
Comment (bmoore-msft) [reply 101-machine-learning-create/azuredeploy.json:438]: It's fine to use an existing storageAccount but you can't modify it in this way, it will remove existing configuration which will break other resources relying on it

File: 101-machine-learning-create/azuredeploy.json (hunk 27 above)
Comment (bmoore-msft) [reply 101-machine-learning-create/azuredeploy.json:623]: this is not resolved

Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:623]: @bmoore-msft I took the dns zone and virtual link out. For the dns records copy, it has to stay there

File: 101-machine-learning-create/azuredeploy.json (hunk 16 above)
Comment (guanyu-240) [reply 101-machine-learning-create/azuredeploy.json:438]: I made one more change on this kind of update.
